power_grid_model_c/power_grid_model/include/power_grid_model/grouped_index_vector.hpp

```diff
@@ -0,0 +1,255 @@
+                                                                    { t - n } -> std::same_as<T>;
+                                                                    { t += n } -> std::same_as<T&>;
+                                                                    { t -= n } -> std::same_as<T&>;
+                                                                };
+
+template <typename T, typename ElementType>
+concept random_access_iterable_like = requires(T const t) {
+                                          { t.begin() } -> random_access_iterator_like<ElementType>;
+                                          { t.end() } -> random_access_iterator_like<ElementType>;
+                                      };
+
+template <typename T>
+concept index_range_iterator =
+    random_access_iterator_like<T, typename T::iterator> && requires(T const t) {
+                                                                typename T::iterator;
+                                                                { *t } -> random_access_iterable_like<Idx>;
+                                                            };
+
+template <typename T>
+concept grouped_index_vector_type = std::default_initializable<T> && requires(T const t, Idx const idx) {
+                                                                         { t.size() } -> std::same_as<Idx>;
+
+                                                                         { t.begin() } -> index_range_iterator;
+                                                                         { t.end() } -> index_range_iterator;
+                                                                         {
+                                                                             t.get_element_range(idx)
+                                                                             } -> random_access_iterable_like<Idx>;
+
+                                                                         { t.element_size() } -> std::same_as<Idx>;
+                                                                         { t.get_group(idx) } -> std::same_as<Idx>;
+                                                                     };
+} // namespace detail
+
+template <typename T>
+concept grouped_idx_vector_type = detail::grouped_index_vector_type<T>;
+
+struct from_sparse_t {};
+struct from_dense_t {};
+
+constexpr auto from_sparse = from_sparse_t{};
+constexpr auto from_dense = from_dense_t{};
+
+class DenseGroupedIdxVector;
+
+class SparseGroupedIdxVector {
+  private:
+    class GroupIterator : public boost::iterator_facade<GroupIterator, Idx, boost::random_access_traversal_tag,
+                                                        boost::iterator_range<IdxCount>, Idx> {
+      public:
+        using iterator = boost::iterator_range<IdxCount>;
+
+        explicit constexpr GroupIterator(IdxVector const& indptr, Idx group) : indptr_(indptr), group_(group) {}
+
+      private:
+        IdxVector const& indptr_;
```

Comment:
Iterator needs to be default constructible and assignable. I think that's the reason you do not satisify the requirements of standard iterator concept.

Please define this as a pointer. The default constructor will assign this pointer to `nullptr`.

Reply:
i did an investigation for this and thought i found that the boost `counting_iterator` does not satisfy the concept, but i will double-check

Reply:
as discussed: i fixed the `nullptr` part. the `counting_iterator` does not satisfy the `std::random_access_iterator` constraints. I added a TODO with the following static assert:

```cpp
// TODO(mgovers): replace the below relevant subset here ourselves with the STD equivalent when we have std::ranges.
// boost::counting_iterator does not satisfy all requirements std::*_iterator concepts:
static_assert(!std::random_access_iterator<IdxCount>);
// we have to declare the relevant subset here ourselves.
```